File: src/lib/tsc/tsc.c
#include "lib/tsc/tsc.h"

#include <assert.h>
#include <cpuid.h>
#include <errno.h>
#include <inttypes.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <x86intrin.h>

#include "lib/logger/logger.h"

// Returns 0 on failure.
static uint64_t _frequency_via_cpuid0x15() {
    unsigned int a = 0, b = 0, c = 0, d = 0;
    // Use the cpuid instruction (wrapped by __get_cpuid) to determine the clock
    // frequency. See "cpuid" in "Intel® 64 and IA-32 Architectures Software
    // Developer’s Manual Volume 2A".

    if (!__get_cpuid(0x0, &a, &b, &c, &d)) {
        // Should never happen.
        panic("cpuid");
    }
    const unsigned int max_level = a;

    if (max_level < 0x15) {
        debug("cpuid 0x15 unsupported; can't get tsc frequency");
        return 0;
    }

    if (!__get_cpuid(0x15, &a, &b, &c, &d)) {
        debug("cpuid 0x15 failed");
        return 0;
    }
    // From "cpuid": "An unsigned integer which is the denominator of the
    // TSC/'core crystal clock' ratio."
    const unsigned int denominator = a;
    if (!denominator) {
        debug("cpuid 0x15 didn't give denominator");
        return 0;
    }
    // From "cpuid": "An unsigned integer which is the numerator of the
    // TSC/'core crystal clock' ratio."
    const unsigned int numerator = b;
    if (!numerator) {
        debug("cpuid 0x15 didn't give numerator");
        return 0;
    }
    // From "cpuid": "An unsigned integer which is the nominal frequency of the
    // core crystal clock in Hz."
    unsigned int core = c;
    if (!core) {
        // From "cpuid": "If ECX is 0, the nominal core crystal clock frequency
        // is not enumerated". Gee, thanks.
        //
        // "Intel® 64 and IA-32 ArchitecturesSoftware Developer’s Manual
        // Volume 3B: System Programming Guide, Part 2", "18.18 COUNTING CLOCKS",
        // gives a 2 row table for this case:
        //
        //   6th and 7th generation Intel® Core™ processors -> 24 MHz
        //
        //   Next Generation Intel® Atom™ processors based on Goldmont
        //   Microarchitecture with CPUID signature 06_5CH -> 19.2 MHz.
        //
        // I wasn't able to find a precise mapping from "06_5CH" to exactly a
        // cpuid result in the Intel manual, but from
        // https://en.wikichip.org/wiki/intel/cpuid, it appears to mean "family
        // 0x6, extended model 0x5, model 0xc", as returned by cpuid 0x1.
        //
        // AFAICT from https://www.amd.com/system/files/TechDocs/25481.pdf, AMD
        // processors don't support cpuid 0x15 at all, so we would've already
        // bailed out earlier for those.
        if (!__get_cpuid(0x1, &a, &b, &c, &d)) {
            debug("cpuid 0x1 failed");
            return 0;
        }
        // bits 11-8
        unsigned int family_id = (a >> 8) & 0xf;
        // bits 19-16
        unsigned int extended_model_id = (a >> 16) & 0xf;
        // bits 7-4
        unsigned int model = (a >> 4) & 0xf;
        trace("rax %u -> family_id:0x%x extended_model_id:0x%x model:0x%x", a, family_id,
              extended_model_id, model);
        if (family_id == 0x6 && extended_model_id == 0x5 && model == 0xc) {
            trace("goldmont; using 19.2 MHz crystal frequency");
            core = 19200000;
        } else {
            trace("non-goldmont; using 24 MHz crystal frequency");
            core = 24000000;
        }
    }

    uint64_t freq = (uint64_t)core * numerator / denominator;
    debug("Calculated %" PRIu64 " cyclesPerSecond via cpuid 15h", freq);
    return freq;
}

// Returns 0 on failure.
static uint64_t _frequency_via_brand_string() {
    // While this *sounds* hacky at first glance, the cpuid docs provide a very
    // precise specification for parsing the cpu frequency out of the brand
    // string.

    unsigned int a = 0, b = 0, c = 0, d = 0;

    if (!__get_cpuid(0x80000000, &a, &b, &c, &d)) {
        debug("cpuid 0x80000000 failed");
        return 0;
    }
    if (!(a & 0x80000000)) {
        // This *shouldn't* happen. The docs say this method is supported on
        // "all Intel 64 and IA-32 processors."
        debug("Brand string method unsupported. Out of fallbacks for getting frequency.");
        return 0;
    }

    // We need to call cpuid 3 times, each time getting 4*4 bytes of the string.
    union {
        uint32_t ints[3 /*calls*/][4 /*regs*/];
        char chars[3 /*calls*/ * 4 /*regs*/ * sizeof(uint32_t)];
    } brand_string;
    for (int i = 0; i < 3; ++i) {
        if (!__get_cpuid(0x80000002 + i, &a, &b, &c, &d)) {
            panic("cpuid");
        }
        brand_string.ints[i][0] = a;
        brand_string.ints[i][1] = b;
        brand_string.ints[i][2] = c;
        brand_string.ints[i][3] = d;
    }
    // Guaranteed to be null terminated.
    assert(brand_string.chars[sizeof(brand_string) - 1] == '\0');

    trace("Got brand string %s", brand_string.chars);

    // Docs say to reverse scan for a 'blank', last token should always be of
    // form x.yz[MGT]Hz.

    char* last_token = rindex(brand_string.chars, ' ');
    assert(last_token);
    assert(*last_token == ' ');

    float base_frequency;
    char scale_c;
    if (sscanf(last_token, " %f%cHz", &base_frequency, &scale_c) != 2) {
        panic("Couldn't parse %s", last_token);
    }
    uint64_t scale = 0;
    if (scale_c == 'M') {
        scale = 1000000ull;
    } else if (scale_c == 'G') {
        scale = 1000000000ull;
    } else if (scale_c == 'T') {
        scale = 1000000000000ull;
    } else {
        panic("Unrecognized scale character %c", scale_c);
    }

    uint64_t frequency = (uint64_t)(base_frequency * scale);
    debug("Calculated %" PRIu64 " cyclesPerSecond via brand string", frequency);
    return frequency;
}

Tsc Tsc_init() {
    // Since we don't have an efficient way of trapping and emulating cpuid
    // to just dictate the perceived clock frequency to the managed program,
    // we need to use cpuid ourselves to figure out the clock frequency, so that
    // we can have the TSC tick at the expected rate when compared to the simulated
    // time retrieved by other means (e.g. clock_gettime).

    uint64_t f = _frequency_via_cpuid0x15();
    if (!f) {
        f = _frequency_via_brand_string();
    }
    if (!f) {
        // If this becomes an issue in practice, we could fall back to measuring
        // empirically (and rounding for attempted determinism?), or just using
        // a fixed constant.
        panic("Couldn't get CPU frequency");
    }

    return (Tsc){.cyclesPerSecond = f};
}

static void _Tsc_setRdtscCycles(const Tsc* tsc, uint64_t* rax, uint64_t* rdx, uint64_t nanos) {
    assert(tsc);
    assert(rax);
    assert(rdx);
    // Guaranteed not to overflow since the operands are both 64 bit.
    __uint128_t gigaCycles = (__uint128_t)tsc->cyclesPerSecond * nanos;
    uint64_t cycles = gigaCycles / 1000000000;
    *rdx = (cycles >> 32) & 0xffffffff;
    *rax = cycles & 0xffffffff;
}

void Tsc_emulateRdtsc(const Tsc* tsc, uint64_t* rax, uint64_t* rdx, uint64_t* rip, uint64_t nanos) {
    assert(tsc);
    assert(rax);
    assert(rdx);
    _Tsc_setRdtscCycles(tsc, rax, rdx, nanos);
    *rip += 2;
}

void Tsc_emulateRdtscp(const Tsc* tsc, uint64_t* rax, uint64_t* rdx, uint64_t* rcx, uint64_t* rip,
                       uint64_t nanos) {
    assert(tsc);
    assert(rax);
    assert(rdx);
    assert(rcx);
    _Tsc_setRdtscCycles(tsc, rax, rdx, nanos);
    // rcx is set to IA32_TSC_AUX. According to the Intel developer manual
    // 17.17.2 "IA32_TSC_AUX Register and RDTSCP Support", "IA32_TSC_AUX
    // provides a 32-bit field that is initialized by privileged software with a
    // signature value (for example, a logical processor ID)." ... "User mode
    // software can use RDTSCP to detect if CPU migration has occurred between
    // successive reads of the TSC. It can also be used to adjust for per-CPU
    // differences in TSC values in a NUMA system."
    //
    // For now we just hard-code an arbitrary constant, which should be fine for
    // the stated purpose.
    // `hex(int(random.random()*2**32))`
    *rcx = 0x806eb479;
    *rip += 3;
}

File: src/lib/tsc/src/tsc.rs
use crate::c_internal;

/// Emulates an x86-64 processor's timestamp counter, as read by rdtsc and
/// rdtscp.
#[repr(C)]
#[allow(non_snake_case)]
pub struct Tsc {
    // TODO: rename and make non-pub when we drop C API
    pub cyclesPerSecond: u64,
}

impl Tsc {
    /// Returns the host system's native TSC rate, or None if it couldn't be found.
    ///
    /// WARNING: this is known to fail completely on some supported CPUs
    /// (particularly AMD), and can return the wrong value for others. i.e. this
    /// needs more work if we need to dependably get the host's TSC rate.
    /// e.g. see <https://github.com/shadow/shadow/issues/1519>.
    pub fn native_cycles_per_second() -> Option<u64> {
        let res = unsafe { c_internal::TscC_nativeCyclesPerSecond() };
        if res == 0 { None } else { Some(res) }
    }

    pub fn new(cycles_per_second: u64) -> Self {
        Self {
            cyclesPerSecond: cycles_per_second,
        }
    }

    fn set_rdtsc_cycles(&self, rax: &mut u64, rdx: &mut u64, nanos: u64) {
        // The multiply is guaranteed not to overflow since both operands are 64 bit.
        let cycles = u128::from(self.cyclesPerSecond) * u128::from(nanos) / 1_000_000_000;
        // *possible* that we'll wrap around here, but only after a very long
        // simulated time and/or a ridiculously fast clock. Wrapping is also
        // presumably what would happen on real hardware.
        let cycles = cycles as u64;
        *rdx = (cycles >> 32) & 0xff_ff_ff_ff;
        *rax = cycles & 0xff_ff_ff_ff;
    }

    const RDTSC: [u8; 2] = [0x0f, 0x31];
    const RDTSCP: [u8; 3] = [0x0f, 0x01, 0xf9];

    /// Updates registers to reflect the result of executing an rdtsc
    /// instruction at time `nanos`.
    pub fn emulate_rdtsc(&self, rax: &mut u64, rdx: &mut u64, rip: &mut u64, nanos: u64) {
        self.set_rdtsc_cycles(rax, rdx, nanos);
        *rip += Self::RDTSC.len() as u64;
    }

    /// Updates registers to reflect the result of executing an rdtscp
    /// instruction at time `nanos`.
    pub fn emulate_rdtscp(
        &self,
        rax: &mut u64,
        rdx: &mut u64,
        rcx: &mut u64,
        rip: &mut u64,
        nanos: u64,
    ) {
        self.set_rdtsc_cycles(rax, rdx, nanos);
        *rip += Self::RDTSCP.len() as u64;

        // rcx is set to IA32_TSC_AUX. According to the Intel developer manual
        // 17.17.2 "IA32_TSC_AUX Register and RDTSCP Support", "IA32_TSC_AUX
        // provides a 32-bit field that is initialized by privileged software with a
        // signature value (for example, a logical processor ID)." ... "User mode
        // software can use RDTSCP to detect if CPU migration has occurred between
        // successive reads of the TSC. It can also be used to adjust for per-CPU
        // differences in TSC values in a NUMA system."
        //
        // For now we just hard-code an arbitrary constant, which should be fine for
        // the stated purpose.
        // `hex(int(random.random()*2**32))`
        *rcx = 0x806eb479;
    }

    /// SAFETY: `ip` must be a dereferenceable pointer, pointing to the beginning
    /// of a valid x86_64 instruction, and `insn` must be a valid x86_64 instruction.
    unsafe fn ip_matches(ip: *const u8, insn: &[u8]) -> bool {
        // SAFETY:
        // * Caller has guaranteed that `ip` points to some valid instruction.
        // * Caller has guaranteed that `insn` is a valid instruction.
        // * No instruction can be a prefix of another, so `insn` can't be a prefix
        //   of some *other* instruction at `ip`.
        // * [`std::Iterator::all`] is short-circuiting.
        //
        // e.g. consider the case where `ip` points to a 1-byte `ret`
        // instruction, and the next byte of memory isn't accessible. That
        // single byte *cannot* match the first byte of `insn`, so we'll never
        // dereference `ip.offset(1)`, which would be unsound.
        insn.iter()
            .enumerate()
            .all(|(offset, byte)| unsafe { *ip.add(offset) == *byte })
    }

    /// Whether `ip` points to an rdtsc instruction.
    ///
    /// # Safety
    ///
    /// `ip` must be a dereferenceable pointer, pointing to the
    /// beginning of a valid x86_64 instruction.
    pub unsafe fn ip_is_rdtsc(ip: *const u8) -> bool {
        unsafe { Self::ip_matches(ip, &Self::RDTSC) }
    }

    /// Whether `ip` points to an rdtscp instruction.
    ///
    /// # Safety
    ///
    /// `ip` must be a dereferenceable pointer, pointing to the
    /// beginning of a valid x86_64 instruction.
    pub unsafe fn ip_is_rdtscp(ip: *const u8) -> bool {
        unsafe { Self::ip_matches(ip, &Self::RDTSCP) }
    }
}

mod export {
    use super::*;

    /// Returns the host system's native TSC rate, or 0 if it couldn't be found.
    ///
    /// WARNING: this is known to fail completely on some supported CPUs
    /// (particularly AMD), and can return the wrong value for others. i.e. this
    /// needs more work if we need to dependably get the host's TSC rate.
    /// e.g. see https://github.com/shadow/shadow/issues/1519.
    #[unsafe(no_mangle)]
    pub extern "C-unwind" fn Tsc_nativeCyclesPerSecond() -> u64 {
        Tsc::native_cycles_per_second().unwrap_or(0)
    }

    /// Instantiate a TSC with the given clock rate.
    #[unsafe(no_mangle)]
    pub extern "C-unwind" fn Tsc_create(cycles_per_second: u64) -> Tsc {
        Tsc::new(cycles_per_second)
    }

    /// Updates `regs` to reflect the result of executing an rdtsc instruction at
    /// time `nanos`.
    #[unsafe(no_mangle)]
    pub extern "C-unwind" fn Tsc_emulateRdtsc(
        tsc: *const Tsc,
        rax: *mut u64,
        rdx: *mut u64,
        rip: *mut u64,
        nanos: u64,
    ) {
        let tsc = unsafe { tsc.as_ref().unwrap() };
        let rax = unsafe { rax.as_mut().unwrap() };
        let rdx = unsafe { rdx.as_mut().unwrap() };
        let rip = unsafe { rip.as_mut().unwrap() };
        tsc.emulate_rdtsc(rax, rdx, rip, nanos)
    }

    /// Updates `regs` to reflect the result of executing an rdtscp instruction at
    /// time `nanos`.
    #[unsafe(no_mangle)]
    pub extern "C-unwind" fn Tsc_emulateRdtscp(
        tsc: *const Tsc,
        rax: *mut u64,
        rdx: *mut u64,
        rcx: *mut u64,
        rip: *mut u64,
        nanos: u64,
    ) {
        let tsc = unsafe { tsc.as_ref().unwrap() };
        let rax = unsafe { rax.as_mut().unwrap() };
        let rdx = unsafe { rdx.as_mut().unwrap() };
        let rcx = unsafe { rcx.as_mut().unwrap() };
        let rip = unsafe { rip.as_mut().unwrap() };
        tsc.emulate_rdtscp(rax, rdx, rcx, rip, nanos)
    }

    /// Whether `buf` begins with an rdtsc instruction.
    #[unsafe(no_mangle)]
    pub extern "C-unwind" fn isRdtsc(ip: *const u8) -> bool {
        unsafe { Tsc::ip_is_rdtsc(ip) }
    }

    /// Whether `buf` begins with an rdtscp instruction.
    #[unsafe(no_mangle)]
    pub extern "C-unwind" fn isRdtscp(ip: *const u8) -> bool {
        unsafe { Tsc::ip_is_rdtscp(ip) }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    fn get_emulated_cycles(clock: u64, nanos: u64) -> u64 {
        let tsc = Tsc::new(clock);

        let mut rax = 0;
        let mut rdx = 0;
        let mut rcx = 0;
        let mut rip = 0;

        tsc.emulate_rdtsc(&mut rax, &mut rdx, &mut rip, nanos);
        assert_eq!(rax >> 32, 0);
        assert_eq!(rdx >> 32, 0);
        let rdtsc_res = (rdx << 32) | rax;

        tsc.emulate_rdtscp(&mut rax, &mut rdx, &mut rcx, &mut rip, nanos);
        assert_eq!(rax >> 32, 0);
        assert_eq!(rdx >> 32, 0);
        let rdtscp_res = (rdx << 32) | rax;

        assert_eq!(rdtsc_res, rdtscp_res);
        rdtsc_res
    }

    #[test]
    fn ns_granularity_at_1_ghz() {
        assert_eq!(get_emulated_cycles(1_000_000_000, 1), 1);
    }

    #[test]
    fn scales_with_clock_rate() {
        let base_clock = 1_000_000_000;
        let base_nanos = 1;
        assert_eq!(
            get_emulated_cycles(1000 * base_clock, base_nanos),
            1000 * get_emulated_cycles(base_clock, base_nanos)
        );
    }

    #[test]
    fn scales_with_time() {
        let base_clock = 1_000_000_000;
        let base_nanos = 1;
        assert_eq!(
            get_emulated_cycles(base_clock, 1000 * base_nanos),
            1000 * get_emulated_cycles(base_clock, base_nanos)
        );
    }

    #[test]
    fn large_cycle_count() {
        let one_year_in_seconds: u64 = 365 * 24 * 60 * 60;
        let ten_b_cycles_per_second: u64 = 10_000_000_000;
        let expected_cycles = one_year_in_seconds
            .checked_mul(ten_b_cycles_per_second)
            .unwrap();
        let actual_cycles =
            get_emulated_cycles(ten_b_cycles_per_second, one_year_in_seconds * 1_000_000_000);
        assert_eq!(actual_cycles, expected_cycles);
    }
}
